lib/client/aws_role/impl.go

```diff
@@ -0,0 +1,213 @@
+
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+const toothlessPolicy = `{
+    "Version": "2012-10-17",
+    "Statement": {
+        "Sid": "HandCrafted",
+        "Effect": "Allow",
+        "Action": "sts:GetCallerIdentity",
+        "Resource": "*"
+    }
+}
+`
+
+func parseArn(arnString string) (*arn.ARN, error) {
+	parsedArn, err := arn.Parse(arnString)
+	if err != nil {
+		return nil, err
+	}
+	switch parsedArn.Service {
+	case "iam", "sts":
+	default:
+		return nil, fmt.Errorf("unsupported service: %s", parsedArn.Service)
+	}
+	splitResource := strings.Split(parsedArn.Resource, "/")
+	if len(splitResource) < 2 || splitResource[0] != "assumed-role" {
+		return nil, fmt.Errorf("invalid resource: %s", parsedArn.Resource)
+	}
+	parsedArn.Service = "iam"
+	parsedArn.Resource = "role/" + splitResource[1]
+	return &parsedArn, nil
+}
+
+func newManager(p Params) (*Manager, error) {
+	cert, err := p.getRoleCertificateTLS()
+	if err != nil {
+		return nil, err
+	}
+	p.Logger.Printf("got AWS Role certificate for: %s\n", p.roleArn)
+	manager := &Manager{
+		Params:  p,
+		tlsCert: cert,
+	}
+	go manager.refreshLoop()
+	return manager, nil
+}
+
+func (m *Manager) getClientCertificate(cri *tls.CertificateRequestInfo) (
+	*tls.Certificate, error) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+	return m.tlsCert, m.tlsError
+}
+
+func (m *Manager) refreshLoop() {
+	for ; ; time.Sleep(time.Minute) {
+		m.refreshOnce()
+	}
+}
+
+func (m *Manager) refreshOnce() {
+	if m.tlsCert != nil {
+		refreshTime := m.tlsCert.Leaf.NotBefore.Add(
+			m.tlsCert.Leaf.NotAfter.Sub(m.tlsCert.Leaf.NotBefore) * 3 / 4)
+		time.Sleep(time.Until(refreshTime))
+	}
+	if cert, err := m.getRoleCertificateTLS(); err != nil {
+		m.Logger.Println(err)
+		if m.tlsCert == nil {
+			m.mutex.Lock()
+			m.tlsError = err
+			m.mutex.Unlock()
+		}
+	} else {
+		m.mutex.Lock()
+		m.tlsCert = cert
+		m.tlsError = nil
+		m.mutex.Unlock()
+		m.Logger.Printf("refreshed AWS Role certificate for: %s\n", m.roleArn)
+	}
+}
+
+// Returns certificate PEM block.
+func (p *Params) getRoleCertificate() ([]byte, error) {
+	if err := p.setupVerify(); err != nil {
+		return nil, err
+	}
+	presignedReq, err := p.stsPresignClient.PresignGetCallerIdentity(p.Context,
+		&sts.GetCallerIdentityInput{})
+	if err != nil {
+		return nil, err
+	}
+	p.Logger.Debugf(1, "presigned URL: %v\n", presignedReq.URL)
+	hostPath := p.KeymasterServer + paths.RequestAwsRoleCertificatePath
+	body := &bytes.Buffer{}
+	body.Write(p.pemPubKey)
+	req, err := http.NewRequestWithContext(p.Context, "POST", hostPath, body)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("claimed-arn", p.roleArn)
+	req.Header.Add("presigned-method", presignedReq.Method)
+	req.Header.Add("presigned-url", presignedReq.URL)
+	resp, err := p.HttpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
```

Comment:
please consume the body of the response even on the non 200 case here.

Reply:
Done.